neo4j-to-neptune/src/main/java/com/amazonaws/services/neptune/util/NeptuneBulkLoader.java

```
@@ -270,66 +273,117 @@ private CompletableFuture<Boolean> uploadFilesSequentially(File[] files, String
+                    } else {
+                        throw new RuntimeException("Upload or compression failed", throwable);
+                    }
+                })
+                .whenComplete((result, throwable) -> {
+                    // Always shutdown the executor service
+                    streamExecutor.shutdown();
+                });
+        }
+    }
+
+    /**
+     * Start the compression task in a background thread
+     */
+    protected CompletableFuture<Void> startCompressionTask(File localFile, PipedOutputStream pipedOut) {
+        return CompletableFuture.runAsync(() -> {
+            try (GZIPOutputStream gzipOut = new GZIPOutputStream(pipedOut);
+                 FileInputStream fis = new FileInputStream(localFile);
+                 BufferedInputStream fileIn = new BufferedInputStream(fis)) {
+
+                // Use same buffer size as GzipCompressUtils for consistency
+                byte[] buffer = new byte[8192];
+                int bytesRead;
+                while ((bytesRead = fileIn.read(buffer)) != -1) {
+                    gzipOut.write(buffer, 0, bytesRead);
                 }
-            });
+                gzipOut.finish();
 
-        } catch (Exception e) {
-            System.err.println("Error initiating Transfer Manager upload for " + localFilePath + ": " + e.getMessage());
-            e.printStackTrace();
-            return CompletableFuture.completedFuture(false);
+            } catch (IOException e) {
+                throw new UncheckedIOException("Compression failed for " + localFile.getAbsolutePath(), e);
+            }
+        });
+    }
+
+    /**
+     * Create the S3 upload request with compressed stream
+     */
+    private UploadRequest createUploadRequest(String s3Key, PipedInputStream pipedIn, ExecutorService streamExecutor) {
+        return UploadRequest.builder()
+            .putObjectRequest(PutObjectRequest.builder()
+                .bucket(bucketName)
```

I don't see any validation of the user input for `bucketName` or `s3Key`. You can add some basic ones using the `com.github.rvesse.airline.annotations` library.

---

What kind of validation are we looking for here? 

The SDK itself will return an error if the bucket does not exist and the s3Key can be anything the user wants as it will just create the folder in the bucket assuming the bucket actually exists in the users account.

---

It's best practice to 'fail early' as to not use resources/take time and make API calls for a workflow that you know will not succeed because of bad input. Similar to how websites will validate your input for text fields and date pickers before allowing requests to the backend API.

For example for s3 buckets there are documented rules for syntax: https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html. It would be easy for us to add simple validation using the CLI library annotations for min and max length and valid characters.

---

Okay added in some basic s3 bucket name validation according to the doc you provided.